### PR TITLE
Addded advanced exaplanation and examples for st.button

### LINKED ISSUE #675: Add more explanations and examples about buttons
Buttons continue to be a common point of frustration for new users. Many times have I said:
>Buttons aren't stateful. They return True on the page load resulting from their click, and then go back to False.

I propose an extensive KB article to point new users to when they nest other widgets inside buttons.

PR #674 created. Opening this issue as notice of work and area of public discussion.

### PR DESCRIPTION
**Pull request by MathCatsAnd**
## 📚 Context
Buttons and what gets nested inside them continues to be a repeat theme in the forum, stack exchange, etc. I feel like a broken record with as many times as I've typed:
> Buttons aren't stateful. They return True on the page load resulting from their click, and then go back to False.

#675 

It would be helpful (and a huge time saver) to have a more robust "button repository" to point to.

## 🧠 Description of Changes
This PR proposes a new KB article all about buttons will a long list of examples showing working and non-working logical patterns. It begins with:
```
## TLDR

Buttons do not retain state. They return `True` on the page load resulting from
their click and then immediately go back to `False`. If you nest something
inside a button, then it will go away as soon as the user takes their next
action (because the page reloads and the button becomes `False`).
```

- [X] Menu/Navigation updated for users to access article
- [ ] Reorganize/re-categorize "Using Streamlit" section of KB to be more accessible/intuitive
- [ ] Host some of the code snippets for interactivity?
- [ ] Update the `st.button` page with link to new KB article as appropriate
- [X] I assume the site map will get updated by itself automatically

## 💥 Impact
Size:
- [X] Not small 

## 🌐 References
This is cheeky, but largely correct: [the forum](https://discuss.streamlit.io/search?q=%40mathcatsand%20buttons)... 😝 
Closes #675

**Contribution License Agreement**
By submitting this pull request you agree that all contributions to this project are made under the Apache 2.0 license.

### CODE REVIEW

**Comment by MathCatsAnd** [comment]
@snehankekre I've marked this as a draft to get some feedback. (@jrieke do you want to be pinged too on these bigger ideas?) I'm not 100% in love with each and every example and I think a couple could be tweaked for clarity, but this is a first pass to get the idea "on paper" and get the ball rolling. 

I'd like to also reorganize the "Using Streamlit" section as the long list of articles doesn't feel very accessible. Maybe most people are getting to those articles through search, but it would be nice to have some grouping/categorization of the articles. I thought it might even be nice to create a separate "widgets" section if there are other widgets that could use a longer-winded explanation and plethora of toy examples beyond their basic API doc page.

<details><summary>Preview</summary>

</details>

**Comment by MathCatsAnd** [comment]
> @MathCatsAnd IIUC the recent feedback from KM has yet to incorporated here. Once that's done, I'll do a final review and merge 😄

Whoops, forgot to push. Done now.

File: content/kb/using-streamlit/index.md [hunk 1]
```diff
@@ -8,6 +8,7 @@ slug: /knowledge-base/using-streamlit
 Here are some frequently asked questions about using Streamlit. If you feel something important is missing that everyone needs to know, please [open an issue](https://github.com/streamlit/docs/issues) or [submit a pull request](https://github.com/streamlit/docs/pulls) and we'll be happy to review it!
 
 - [Sanity checks](/knowledge-base/using-streamlit/sanity-checks)
+- [Buttons, buttons, buttons!](/knowledge-base/using-streamlit/buttons)
```
Comment (snehankekre) [inline content/kb/using-streamlit/index.md:11]: Title here should ideally match that of the article.

File: content/menu.md [hunk 2]
```diff
@@ -473,6 +473,9 @@ site_menu:
   - category: Knowledge base / Using Streamlit / Batch elements and input widgets with st.form
     url: /knowledge-base/using-streamlit/batch-elements-input-widgets-form
     visible: false
+  - category: Knowledge base / Using Streamlit / Buttons, buttons, buttons!
```
Comment (snehankekre) [inline content/menu.md:476]: Title here should ideally match that of the article.

File: content/kb/using-streamlit/buttons.md [hunk 3]
```diff
@@ -0,0 +1,412 @@
+---
+title: Button behavior and examples
+slug: /knowledge-base/using-streamlit/buttons
```
Comment (snehankekre) [inline content/kb/using-streamlit/buttons.md:3]:
```suggestion
slug: /knowledge-base/using-streamlit/button-behavior-and-examples
```

Better this way for SEO.

File: content/kb/using-streamlit/index.md (hunk 1 above)
Comment (snehankekre) [inline content/kb/using-streamlit/index.md:11]: Edit slug to: /knowledge-base/using-streamlit/button-behavior-and-examples

File: content/menu.md [hunk 4]
```diff
@@ -473,6 +473,9 @@ site_menu:
...
+    url: /knowledge-base/using-streamlit/buttons
```
Comment (snehankekre) [inline content/menu.md:477]: Same suggestion as above for the slug :)

File: content/kb/using-streamlit/buttons.md [hunk 5]
```diff
@@ -0,0 +1,412 @@
...
+their click and immediately return to `False` on the next load. If a displayed
+element is nested inside `if st.button('Click me'):`, the element will be
+visible when the button is clicked and disappear as soon as the user takes their
+next action. This is because the page reloads and the button becomes `False`.
+
+## Use cases for buttons
+
+When code is conditioned on a button's value, it will execute once in
+response to the button being clicked and not again (until the button is clicked
+again).
```
Comment (snehankekre) [inline content/kb/using-streamlit/buttons.md:20]: Perhaps reword slightly or add some more text here? It's unclear why this sentence follows a heading that says use cases.

File: content/kb/using-streamlit/buttons.md [hunk 6]
```diff
@@ -0,0 +1,412 @@
...
+
+### Nest inside of a button:
```
Comment (snehankekre) [inline content/kb/using-streamlit/buttons.md:22]: Nest what? 😅

File: content/kb/using-streamlit/buttons.md [hunk 7]
```diff
@@ -0,0 +1,412 @@
...
+
+- Transient messages that immediately disappear.
+- Once-per-click processes that saves data to session state, a file, or
+  a database.
+
+### Do not nest inside of a button:
```
Comment (snehankekre) [inline content/kb/using-streamlit/buttons.md:28]: Given there's no preamble before these two headings, I find it difficult to understand how these two headings are use cases. We need to flesh this section out a bit more for clarity.

File: content/kb/using-streamlit/buttons.md [hunk 8]
```diff
@@ -0,0 +1,412 @@
...
+animal_shelter = ['cat', 'dog', 'rabbit', 'bird']
+
+animal = st.text_input('Type an animal')
+
+if st.button('Check availability') and animal.lower() in animal_shelter:
+    # This goes away if the user clicks the other button or changes the animal
+    st.write('We have that animal!')
+
+st.button('OK')
+```
```
Comment (snehankekre) [inline content/kb/using-streamlit/buttons.md:58]: We should expand on what this code does and how it relates to the preceding sentence. "In the above/below example, ...."

File: content/kb/using-streamlit/buttons.md [hunk 9]
```diff
@@ -0,0 +1,412 @@
...
+
+### Stateful button
+
+If you want a clicked button to continue to be `True`, create a value in
+`st.session_state` and use the button set it to `True` in a callback:
```
Comment (snehankekre) [inline content/kb/using-streamlit/buttons.md:63]: hyperlink to `/library/advanced-features/session-state`

File: content/library/advanced-features/button-behavior-and-examples.md [hunk 10]
```diff
@@ -0,0 +1,418 @@
...
+if st.button('Process'):
+    result = expensive_process(option, add)
+    st.session_state.processed[option] = result
+
+if option in st.session_state.processed:
+    st.write(f'Option {option} processed with add {add}')
+    st.write(st.session_state.processed[option][0])
+```
+
+Astute observers may think, "This feels a little like caching." We are only saving results relative to one parameter, but the pattern could easily be expanded to save results relative to both parameters. In that sense, yes, it has some similarities to caching, but also some important differences. When you save results in `st.session_state`, the results are only available to the current user in their current session. If you use [`st.cache_data`](/library/api-reference/performance/st.cache_data) instead, the results are available to all users across all sessions. Furthermore, if you want to update a saved result, you have to clear all saved results for that function to do so.
```
Comment (sfc-gh-dmatthews) [inline content/library/advanced-features/button-behavior-and-examples.md:371]: Am I being too obtuse here?

**Review by snehankekre — APPROVED** [review]
